pymc3/gp/cov.py

```diff
@@ -97,29 +102,69 @@ def __init__(self, factor_list):
             else:
                 self.factor_list.append(factor)
 
-    def merge_factors(self, X, Z=None, diag=False):
+    def merge_factors(self, X, Xs=None, diag=False):
+        # this function makes sure diag=True is handled properly
         factor_list = []
         for factor in self.factor_list:
+
+            # if factor is a Covariance
             if isinstance(factor, Covariance):
-                factor_list.append(factor(X, Z, diag))
-            elif hasattr(factor, "ndim"):
-                if diag:
-                    factor_list.append(tt.diag(factor))
-                else:
-                    factor_list.append(factor)
-            else:
-                factor_list.append(factor)
+                factor_list.append(factor(X, Xs, diag))
+                continue
+
+            # if factor is a numpy array
+            if isinstance(factor, np.ndarray):
+                if np.ndim(factor) == 2:
+                    if diag:
+                        factor_list.append(np.diag(factor))
+                        continue
+
+            # if factor is a theano variable with ndim attribute
+            if isinstance(factor, (tt.TensorConstant,
+                                     tt.TensorVariable,
+                                     tt.sharedvar.TensorSharedVariable)):
+                if factor.ndim == 2:
+                    if diag:
+                        factor_list.append(tt.diag(factor))
+                        continue
+
+            # othewise
+            factor_list.append(factor)
+
         return factor_list
 
 
 class Add(Combination):
-    def __call__(self, X, Z=None, diag=False):
-        return reduce((lambda x, y: x + y), self.merge_factors(X, Z, diag))
+    def __call__(self, X, Xs=None, diag=False):
+        return reduce((lambda x, y: x + y), self.merge_factors(X, Xs, diag))
```

Comment:
I'll take a look again.  I remember it was tricky but don't remember why exactly.